oeps/oep-38-Data-Modeling.rst

```diff
@@ -0,0 +1,442 @@
+reconstruct the data at different points with much greater reliability.
+
+If, for some reason, it is not possible to create history at the time of
+entry (for example, if we expect the table to become too large or the
+writes to be too frequent), We recommend that an event be issued in its
+place.
+
+Fire events for CUD operations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We highly recommend firing events for any create, update or delete (CUD)
+operations that happen within the OpenedX ecosystem. The cost of firing
+and storing these events is low, and will allow teams in the future to
+reconstruct the state of the application at various times.
+
+Data Modeling Standards
+-----------------------
+
+All models in the edX ecosystem should have:
+
+-  A Primary Key
+
+   -  Should be named: “id”
+
+   -  Use numeric based identifiers
+
+   -  A key size large enough so the keys will not run out for a very
+         long time.
+
+   -  Do not use composite based primary keys. Use a primary key column
+
+-  Updated and created timestamps
+
+   -  Use UTC time.
+
+   -  Created date should be named: “created_date”
+
+   -  Updated date should be named: “modified_date”
+
+-  Use a numeric foreign key(s)
+
+   -  Data should be joined on other foreign keys
+
+   -  Do not join on things such as username, email address, or other
+         dimensions of data that may change over time.
+
+-  The minimum number of indexes possible to make the table/queries
+      performant
+
+   -  Indexes cost space and have their own set of performance concerns.
+
+   -  Over-indexing data could actually make the database less
+         performant (slower writes/updates)
+
+   -  You should use as many indexes as you need to make the application
+         performant and no more.
+
+-  History for models involved with enrollments, or anything that may
+      have an effect on financial payments or transactions.
+
+-  History using django-simple-history
+
+   -  Should capture the ID of the user who made the change
+
+-  Where Django simple history is not an option, the following data
+      should be captured:
+
+   -  Fields that were changed
+
+   -  Date & time of the change
+
+   -  ID of the user who initiated the change
+
+-  The correct data type for a column.
+
+   -  Don’t use a long when an Int or bool would do.
+
+   -  Use Integers for your keys
+
+   -  Don’t store an int as varchar
+
+-  Don’t mix types in a column
+
+-  Annotations
+
+   -  All models in the OpenedX ecosystem should be tagged for PII
+      using `code
+      annotations <https://github.com/edx/code-annotations>`__\  following \ `OEP-30 <https://github.com/edx/open-edx-proposals/blob/master/oeps/oep-0030-arch-pii-markup-and-auditing.rst>`__
+
+Best Practices 
+~~~~~~~~~~~~~~~
+
+Deleting data:
+
+-  It is better to have a column to mark the record as inactive than to
+      remove the data from the system using the SQL delete keyword.
```

Comment:
Also note the impact of deletions on data in other models.   Models that use ForeignKey with on_delete=models.CASCADE will lose their values when the object the foreign key points to is deleted.

Reply:
Can we also have a recommendation for how to access only active records by default.  Whether that is django object managers that default filter the active field or managing database views that select only the active fields or something.

Reply:
Is it worth recommending having a `deleted_date` for soft deletes in addition to the flag?